### PR TITLE
replace deprecated golang/gomock with uber/gomock

### LINKED ISSUE #5440: Migrate away from unmaintained `golang/mock`
### Proposal

While working on https://github.com/kedacore/keda/pull/5391 and go generics, I realise that `gomock` is no longer maintained as per [official doc](https://github.com/golang/mock?tab=readme-ov-file). It does not support mocking generics too (discussion [link](https://kubernetes.slack.com/archives/C01JGDP8MB8/p1706481936812209?thread_ts=1705333570.247929&cid=C01JGDP8MB8), gh [issue](https://github.com/golang/mock/issues/621#issuecomment-1477288401))
```
Update, June 2023: This repo and tool are no longer maintained. Please see [go.uber.org/mock](https://github.com/uber/mock) for a maintained fork instead
```

We should migrate away from `gomock` to something else, probably [go.uber.org/mock](https://github.com/uber/mock). I can take a look at this issue this week if I have time

### Use-Case

_No response_

### Is this a feature you are interested in implementing yourself?

Yes

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by rahulii**
_Provide a description of what has been changed_

### Checklist

- [N/A] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [N/A] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [N/A] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Fixes #5440 

Relates to #

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -63,6 +63,7 @@ Here is an overview of all new **experimental** features:
 ### Improvements
 
 - **General**: Add OPENTELEMETRY flag in e2e test YAML ([#5375](https://github.com/kedacore/keda/issues/5375))
+- **General**: Migrate away from unmaintained golang/mock and use uber/gomock ([#5440](https://github.com/kedacore/keda/issues/5440))
```
Comment (dttung2905) [inline CHANGELOG.md:66]: Should we move this to `Others` section ? Wdyt @zroubalik

File: pkg/metricsservice/api/metrics.pb.go [hunk 2]
```diff
@@ -15,8 +15,8 @@
 
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
-// 	protoc-gen-go v1.32.0
```
Comment (dttung2905) [inline pkg/metricsservice/api/metrics.pb.go:18]: Could you help to double check this one for me ? Seems like we are dowgrading the version :thinking:

File: CHANGELOG.md (hunk 1 above)
Comment (zroubalik) [reply CHANGELOG.md:66]: Yes!

Comment (rahulii) [reply CHANGELOG.md]: done, PTAL!

File: hack/tools.go [hunk 3]
```diff
@@ -7,6 +7,7 @@ package tools
 import (
 	// Import code-generator to use in build tools
 	_ "github.com/golang/mock/mockgen"
```
Comment (JorTurFer) [inline hack/tools.go:9]: Can we remove this?

**Review by JorTurFer — COMMENTED** [review]
LGTM! just small nit inline

File: hack/tools.go (hunk 3 above)
Comment (rahulii) [reply hack/tools.go]: yes, also removed it from vendor, ptal!

**Comment by JorTurFer** [comment]
Could you solve the changelog conflict?

**Comment by rahulii** [comment]
> Could you solve the changelog conflict?

done!

**Review by JorTurFer — APPROVED** [review]
LGTM! Thanks for the contribution ❤️

**Comment by JorTurFer** [comment]
> /run-e2e

Only us can trigger the e2e tests because they use real infra 😄

**Review by zroubalik — APPROVED** [review]

**Pull request merged**